File: csharp/tests/TestLinkedRecords.cs
using MyAirtable;
using Xunit;

namespace MyAirtable.Tests;

/// <summary>
/// CS6.2 — Linked records as raw record-ID lists (no wrapper type). Resolution happens by passing
/// the IDs to the linked table's <c>GetAsync</c>. Parity with the Java/Kotlin/Swift/Rust suites.
/// </summary>
public class TestLinkedRecords
{
    private readonly Airtable _airtable = TestSetup.MakeAirtable();

    [Fact]
    public async Task LinkedRecordFieldsRoundTripViaRecordIds()
    {
        var suite = TestSetup.PrimaryKey("Linked", "RoundTrip");
        string? sec1Id = null;
        string? sec2Id = null;
        string? primId = null;
        try
        {
            var sec1 = await _airtable.Secondary.CreateAsync(
                new SecondaryModel { Name = suite + " S1" }
            );
            sec1Id = sec1.Id;
            var sec2 = await _airtable.Secondary.CreateAsync(
                new SecondaryModel { Name = suite + " S2" }
            );
            sec2Id = sec2.Id;

            var prim = await _airtable.Primary.CreateAsync(
                new PrimaryModel
                {
                    PrimaryKey = suite,
                    LinkSingle = new List<string> { sec1Id! },
                    LinkMultiple = new List<string> { sec1Id!, sec2Id! },
                }
            );
            primId = prim.Id;
            Assert.Equal(new List<string> { sec1Id! }, prim.LinkSingle);
            Assert.Equal(new List<string> { sec1Id!, sec2Id! }, prim.LinkMultiple);

            var fetched = await _airtable.Primary.GetAsync(primId!);
            Assert.Equal(new List<string> { sec1Id! }, fetched.LinkSingle);
            Assert.Equal(new List<string> { sec1Id!, sec2Id! }, fetched.LinkMultiple);

            fetched.LinkSingle = new List<string> { sec2Id! };
            fetched.LinkMultiple = new List<string> { sec1Id! };
            var updated = await _airtable.Primary.UpdateAsync(fetched);
            Assert.Equal(new List<string> { sec2Id! }, updated.LinkSingle);
            Assert.Equal(new List<string> { sec1Id! }, updated.LinkMultiple);
        }
        finally
        {
            await TryDeletePrimaryAsync(primId);
            await TryDeleteSecondariesAsync(sec1Id, sec2Id);
        }
    }

    [Fact]
    public async Task LinkedSingleRecordResolvesViaSecondaryGet()
    {
        var suite = TestSetup.PrimaryKey("Linked", "Single");
        string? secId = null;
        string? primId = null;
        try
        {
            var sec = await _airtable.Secondary.CreateAsync(
                new SecondaryModel { Name = suite + " Target", Value = "sv" }
            );
            secId = sec.Id;

            var prim = await _airtable.Primary.CreateAsync(
                new PrimaryModel
                {
                    PrimaryKey = suite,
                    LinkSingle = new List<string> { secId! },
                }
            );
            primId = prim.Id;
            Assert.Single(prim.LinkSingle!);
            var linkedId = prim.LinkSingle![0];
            Assert.Equal(secId, linkedId);

            var linked = await _airtable.Secondary.GetAsync(linkedId);
            Assert.Equal(suite + " Target", linked.Name);
            Assert.Equal("sv", linked.Value);
        }
        finally
        {
            await TryDeletePrimaryAsync(primId);
            await TryDeleteSecondariesAsync(secId);
        }
    }

    [Fact]
    public async Task EmptyLinkSingleDecodesAsNull()
    {
        var suite = TestSetup.PrimaryKey("Linked", "FetchNil");
        var prim = await _airtable.Primary.CreateAsync(
            new PrimaryModel { PrimaryKey = suite + " No Links" }
        );
        var primId = prim.Id;
        try
        {
            // Airtable omits empty link fields entirely; linkSingle decodes as null/empty.
            Assert.True(
                prim.LinkSingle is null or { Count: 0 },
                "empty link field decodes as null/empty"
            );
        }
        finally
        {
            await TryDeletePrimaryAsync(primId);
        }
    }

    [Fact]
    public async Task LinkedMultiRecordsResolveViaSecondaryGet()
    {
        var suite = TestSetup.PrimaryKey("Linked", "Multi");
        string? sec1Id = null;
        string? sec2Id = null;
        string? primId = null;
        try
        {
            var sec1 = await _airtable.Secondary.CreateAsync(
                new SecondaryModel { Name = suite + " T1" }
            );
            sec1Id = sec1.Id;
            var sec2 = await _airtable.Secondary.CreateAsync(
                new SecondaryModel { Name = suite + " T2" }
            );
            sec2Id = sec2.Id;

            var prim = await _airtable.Primary.CreateAsync(
                new PrimaryModel
                {
                    PrimaryKey = suite,
                    LinkMultiple = new List<string> { sec1Id!, sec2Id! },
                }
            );
            primId = prim.Id;

            var linked = await _airtable.Secondary.GetAsync(
                prim.LinkMultiple ?? new List<string>()
            );
            Assert.Equal(2, linked.Count);
            var names = linked.Select(m => m.Name).ToList();
            Assert.Contains(suite + " T1", names);
            Assert.Contains(suite + " T2", names);
        }
        finally
        {
            await TryDeletePrimaryAsync(primId);
            await TryDeleteSecondariesAsync(sec1Id, sec2Id);
        }
    }

    // ---- best-effort cleanup helpers (skip null ids; swallow cleanup errors) ----

    private async Task TryDeletePrimaryAsync(string? primId)
    {
        if (string.IsNullOrEmpty(primId))
            return;
        try
        {
            await _airtable.Primary.DeleteAsync(primId);
        }
        catch (AirtableException)
        {
            // best-effort cleanup — surface the original failure
        }
    }

    private async Task TryDeleteSecondariesAsync(params string?[] secIds)
    {
        var ids = secIds.Where(id => !string.IsNullOrEmpty(id)).Select(id => id!).ToList();
        if (ids.Count == 0)
            return;
        try
        {
            await _airtable.Secondary.DeleteAsync(ids);
        }
        catch (AirtableException)
        {
            // best-effort cleanup — surface the original failure
        }
    }
}

File: cpp/tests/test_linked_records.cpp
// Linked records as raw record-ID lists; resolution via the linked table's
// get() (C# TestLinkedRecords parity, 4 cases).
#include <catch2/catch_test_macros.hpp>

#include <algorithm>

#include "test_setup.hpp"

using namespace myairtable;
using myairtable_tests::make_airtable;
using myairtable_tests::primary_key;

namespace {
void try_remove_primary(Airtable& airtable, const std::optional<std::string>& id) {
    if (!id.has_value() || id->empty()) {
        return;
    }
    try {
        airtable.primary().remove(*id);
    } catch (const AirtableException&) {
    }
}

void try_remove_secondaries(Airtable& airtable,
                            const std::vector<std::optional<std::string>>& ids) {
    for (const auto& id : ids) {
        if (!id.has_value() || id->empty()) {
            continue;
        }
        try {
            airtable.secondary().remove(*id);
        } catch (const AirtableException&) {
        }
    }
}
} // namespace

TEST_CASE("linked: record fields round trip via record ids", "[crud][linked]") {
    auto airtable = make_airtable();
    const auto suite = primary_key("Linked", "RoundTrip");
    std::optional<std::string> sec1_id, sec2_id, prim_id;
    try {
        sec1_id = *airtable.secondary().create(SecondaryModel{.name = suite + " S1"}).id;
        sec2_id = *airtable.secondary().create(SecondaryModel{.name = suite + " S2"}).id;

        auto prim = airtable.primary().create(PrimaryModel{
            .primary_key = suite,
            .link_multiple = std::vector<std::string>{*sec1_id, *sec2_id},
            .link_single = std::vector<std::string>{*sec1_id},
        });
        prim_id = prim.id;
        REQUIRE(prim.link_single == std::vector<std::string>{*sec1_id});
        REQUIRE(prim.link_multiple == std::vector<std::string>{*sec1_id, *sec2_id});

        auto fetched = airtable.primary().get(*prim_id);
        REQUIRE(fetched.link_single == std::vector<std::string>{*sec1_id});
        REQUIRE(fetched.link_multiple == std::vector<std::string>{*sec1_id, *sec2_id});

        fetched.link_single = std::vector<std::string>{*sec2_id};
        fetched.link_multiple = std::vector<std::string>{*sec1_id};
        auto updated = airtable.primary().update(fetched);
        REQUIRE(updated.link_single == std::vector<std::string>{*sec2_id});
        REQUIRE(updated.link_multiple == std::vector<std::string>{*sec1_id});
    } catch (...) {
        try_remove_primary(airtable, prim_id);
        try_remove_secondaries(airtable, {sec1_id, sec2_id});
        throw;
    }
    try_remove_primary(airtable, prim_id);
    try_remove_secondaries(airtable, {sec1_id, sec2_id});
}

TEST_CASE("linked: single record resolves via secondary get", "[crud][linked]") {
    auto airtable = make_airtable();
    const auto suite = primary_key("Linked", "Single");
    std::optional<std::string> sec_id, prim_id;
    try {
        sec_id = *airtable.secondary()
                      .create(SecondaryModel{.name = suite + " Target", .value = "sv"})
                      .id;

        auto prim = airtable.primary().create(PrimaryModel{
            .primary_key = suite,
            .link_single = std::vector<std::string>{*sec_id},
        });
        prim_id = prim.id;
        REQUIRE(prim.link_single->size() == 1);
        const auto linked_id = prim.link_single->front();
        REQUIRE(linked_id == *sec_id);

        auto linked = airtable.secondary().get(linked_id);
        REQUIRE(linked.name == suite + " Target");
        REQUIRE(linked.value == "sv");
    } catch (...) {
        try_remove_primary(airtable, prim_id);
        try_remove_secondaries(airtable, {sec_id});
        throw;
    }
    try_remove_primary(airtable, prim_id);
    try_remove_secondaries(airtable, {sec_id});
}

TEST_CASE("linked: empty link-single decodes as nullopt", "[crud][linked]") {
    auto airtable = make_airtable();
    const auto suite = primary_key("Linked", "FetchNil");
    auto prim = airtable.primary().create(PrimaryModel{.primary_key = suite + " No Links"});
    const auto prim_id = prim.id;
    try {
        // Airtable omits empty link fields entirely; link_single decodes as nullopt/empty.
        REQUIRE((!prim.link_single.has_value() || prim.link_single->empty()));
    } catch (...) {
        try_remove_primary(airtable, prim_id);
        throw;
    }
    try_remove_primary(airtable, prim_id);
}

TEST_CASE("linked: multi records resolve via secondary get", "[crud][linked]") {
    auto airtable = make_airtable();
    const auto suite = primary_key("Linked", "Multi");
    std::optional<std::string> sec1_id, sec2_id, prim_id;
    try {
        sec1_id = *airtable.secondary().create(SecondaryModel{.name = suite + " T1"}).id;
        sec2_id = *airtable.secondary().create(SecondaryModel{.name = suite + " T2"}).id;

        auto prim = airtable.primary().create(PrimaryModel{
            .primary_key = suite,
            .link_multiple = std::vector<std::string>{*sec1_id, *sec2_id},
        });
        prim_id = prim.id;

        auto linked =
            airtable.secondary().get(prim.link_multiple.value_or(std::vector<std::string>{}));
        REQUIRE(linked.size() == 2);
        std::vector<std::string> names;
        for (const auto& model : linked) {
            names.push_back(model.name.value_or(""));
        }
        REQUIRE(std::find(names.begin(), names.end(), suite + " T1") != names.end());
        REQUIRE(std::find(names.begin(), names.end(), suite + " T2") != names.end());
    } catch (...) {
        try_remove_primary(airtable, prim_id);
        try_remove_secondaries(airtable, {sec1_id, sec2_id});
        throw;
    }
    try_remove_primary(airtable, prim_id);
    try_remove_secondaries(airtable, {sec1_id, sec2_id});
}
